pyrefly/lib/alt/class/dataclass.rs

```diff
@@ -180,6 +187,314 @@ impl<'a, Ans: LookupAnswer> AnswersSolver<'a, Ans> {
         Some(ClassSynthesizedFields::new(fields))
     }
 
+    pub fn call_dataclasses_replace(
+        &self,
+        args: &[Expr],
+        keywords: &[Keyword],
+        range: TextRange,
+        errors: &ErrorCollector,
+    ) -> Type {
+        let call_args = args.map(CallArg::expr_maybe_starred);
+        let call_keywords = keywords.map(CallKeyword::new);
+        let obj_name = Name::new_static("obj");
+
+        let obj_ty = match args.first() {
+            Some(first) => self.expr_infer(first, errors),
+            None => Type::Any(AnyStyle::Implicit),
+        };
+        let mut dataclass_types = SmallSet::new();
+        let mut saw_any = false;
+        let mut saw_non_dataclass = false;
+        self.distribute_over_union(&obj_ty, |ty| {
```

Comment:
I think you can simplify this function a lot by moving the core logic to be inside `distribute_over_union`, rather than first collecting the dataclass types and then trying to handle the union. I.e., do something like this:
```
self.distribute_over_union(&obj_ty, |ty| {
  // 1. Check if `ty` is a dataclass.
  // 2. If it is, construct and call a `replace` function like you do below.
}
```
Then I believe you can delete all the code that deals with checking the number of overloads, finding common keywords, etc.